### DISCUSSION TITLE
Connection refused to port 8675, RPi

### DISCUSSION THREAD

**Discussion by maxEntropyProd** [jasonacox/Powerwall-Dashboard#55] [Q&A]
Hi
setup.sh seemed to work fine while installing PD on a RPi4; however, when I pull up Powerwall Power Flow dashboard, the Power Flow window shows "Unable to connect" to localhost:8675. If I point a browser to that port, I get a similar message about site can't be reached.
Any suggestions? Note, my powerwall+ and solar was just installed, but the inverter has some issues and so the system is in standby as they try to troubleshoot that. However, if I log into the gateway it shows the Tesla graphic with power being supplied by the grid. So that's working.
Thanks!

**Comment by jasonacox** [comment]
A few things to try:

1. Ensure you do not have another browser window/tab open that is showing the animation. Chrome and other browsers prevent the iframe from rendering if it opened more than once (I would like someone to explain this to me at some point :)
2. Several of the browsers do not like iframes on "localhost" so try to use the IP address instead (e.g. http://10.0.1.2:9000/).
3. Try to view the animation directly from the pyPowerwall proxy by going to this: http://10.0.1.2:8675/ - also make sure the latest pypowerwall was installed (e.g. v0.5.1): http://10.0.1.2:8675/stats (of course replace the IP address with your RPi)

**Comment by maxEntropyProd** [comment]
hi @jasonacox, thanks for the fast reply! I did post this under discussion Q&A, but will keep it here. Interesting, 10.0.1.2 didn't work (I thought that was the same as localhost); however, if I use the IP of the machine on the local network (192.168...), I still get the same response. Likewise, http://192.168.0.80:8675/stats also shows Unable to Connect. I've closed other open tabs first. It's like there is a firewall on port 8675, but I'm not running any firewall on the RPi

I see you've move this here (discussion) now, so will respond here.

**Reply by maxEntropyProd** [reply]
Does Powerall-Dashboard setup.sh install pyPowerwall? I just assumed it did. if it doesn't, then that's my problem.

**Reply by jasonacox** [reply]
It does, but if you are not able to see the pypowerwall stats page at the URL (or http://localhost:8675/stats), that means it is not running.  Troubleshooting steps:

```bash
# See if pypowerwall gives us any indication why it is not starting
docker logs pypowerwall

# Force a restart of the docker compose stack
 docker-compose -f powerwall.yml up -d
```
Example - on my RPi4
<img width="454" alt="image" src="https://user-images.githubusercontent.com/836718/180895551-f2578add-6754-46fb-b184-170fc9c25a0a.png">

**Reply by maxEntropyProd** [reply]
Ya, I get a bunch of these errors in the log:
`Current thread 0xb6f3c390 (most recent call first):
<no Python frame>
Fatal Python error: pyinit_main: can't initialize time
Python runtime state: core initialized
PermissionError: [Errno 1] Operation not permitted
`
The restart seem good:
`influxdb is up-to-date
pypowerwall is up-to-date
grafana is up-to-date
telegraf is up-to-date
`

**Reply by maxEntropyProd** [reply]
The log seems strange. I noticed the default python is 2.7, but 3.7 is installed. Before you answered I did do a pip3 install on pyPowerwall, then ran 
python3 -m pypowerwall scan
which did find the gateway URL correctly.
Anyway, got to do some other things at the moment, but will get back to this later. Powerwall-Dashboard does look amazing, so I'm looking forward go getting that running.

**Reply by jasonacox** [reply]
The docker container for pypowerwall is using Python 3.8.  What version of Raspbian are you running (`cat /etc/os-release`)?  My test system is Raspbian GNU/Linux 11 (bullseye).

From what I can tell, that error is related to an older libseccomp2 library.  It would be interesting to see if an RPi update would fix this:

```
# upgrade RPi Packages
sudo apt update
sudo apt upgrade

# Reboot RPi for fresh start
```

I'll keep search in case others have seen this same error.

**Reply by maxEntropyProd** [reply]
Ya, I'm running Raspbian GNU/Linux 10 (buster). Against my better judgement, I had updated and upgraded packages before installing PD, and it's showing all packages up to date, at least for Linux 10 buster.

If I run the simple program on the pyPowerwall GitHub page that queries the gateway for information, it correctly pulls all the data from the gateway, so pyPowerwall is working correctly it appears; however, this is for python3. I'm still wondering if there is some issue with python2 vs 3, although that probably wouldn't be the issue with the port. 
I do notice when I try to connect a browser to port 8675 it first gives the error like when you connect to http instead of https with a browser and it complains about lack of security, but then it quickly display the unable to connect page.

**Reply by jasonacox** [reply]
I think I found the issue and it is related to the legacy version of Raspbian you are running (buster) and the new Docker +  Ubuntu/Alpine containers.  I suspect you are probably not the only person running this version so I should create more troubleshooting help!  Thanks for finding this and raising the issue.

## Some Research

This error shows up for several python projects using docker running on Raspbian Buster (as I discovered via a google search).   See https://blog.samcater.com/fix-workaround-rpi4-docker-libseccomp2-docker-20/ and also https://docs.linuxserver.io/faq in section "My host is incompatible with images based on Ubuntu Focal and Alpine 3.13 and later. This only affects 32 bit installs of distros based on Debian Buster.

As to not getting any updates when you `apt upgrade`, it seems Buster version is considered legacy so not all packages are being ported which includes the fix to the libseccomp2 bug causing this error.  

## The Fix?

Obviously, the first fix would be to upgrade your OS to the latest Raspbian (bullseye) but I understand that may not be possible and is a bit involved.  I would still recommend this at some point as Buster is legacy.

There are several other ways to get the fixed libseccomp2 installed. The best approach seems to be using the backports (where the libseccomp2 code is backported to legacy OS versions like Buster):

```bash
# Get signing keys to verify the new packages
sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 04EE7237B7D453EC 648ACFD622F3D138

# Add the Buster backport repository to apt sources.list
echo 'deb http://httpredir.debian.org/debian buster-backports main contrib non-free' | sudo tee -a /etc/apt/sources.list.d/debian-backports.list

sudo apt update
sudo apt install libseccomp2 -t buster-backports

# I recommend you restart and see if that corrects the problem
```

## Alternatives

As you discovered, you can run pypowerwall module directly (not in a docker container) without any problems.  This is because the libseccomp2 is related to the docker container not python specifically.  You could build your own container that doesn't use the Alpine distro and would likely not trip on the libseccomp2 bug:

See https://github.com/jasonacox/pypowerwall/tree/main/proxy#build-your-own

```bash
# clone repo
git clone https://github.com/jasonacox/pypowerwall.git
cd pypowerwall/proxy
```
Edit Dockerfile and remove "-alpine"
```bash
FROM python:3.8
WORKDIR /app
RUN pip3 install pypowerwall==0.5.1 bs4
COPY . .
CMD ["python3", "server.py"]
EXPOSE 8675 
```

```
# Build for local architecture  
docker build -t pypowerwall:latest .

# Run docker
docker run \
-d \
-p 8675:8675 \
-e PW_PORT='8675' \
-e PW_PASSWORD='password' \
-e PW_EMAIL='email@example.com' \
-e PW_HOST='localhost' \
-e PW_TIMEZONE='America/Los_Angeles' \
-e PW_CACHE_EXPIRE='5' \
-e PW_DEBUG='no' \
-e PW_HTTPS='no' \
--name pypowerwall \
--restart unless-stopped \
pypowerwall
```

**Reply by maxEntropyProd** [reply]
Thanks @jasonacox that's some good progress. Here's what happend.

Trying the first option I got:

```
sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 04EE7237B7D453EC 648ACFD622F3D138
Executing: /tmp/apt-key-gpghome.HFgsxFUGGO/gpg.1.sh --keyserver keyserver.ubuntu.com --recv-keys 04EE7237B7D453EC 648ACFD622F3D138
gpg: keyserver receive failed: Server indicated a failure
```

So, that looks like a failure, so I did not proceed further with option 1.

Second option:
I got to "docker build -t pypowerwall:latest ." step, but that build kicked out with error:

```
Sending build context to Docker daemon  50.18kB
Step 1/6 : FROM python:3.8
3.8: Pulling from library/python
a8a55ed160b9: Already exists 
5042ae471fd5: Already exists 
70da2acc639b: Already exists 
34f084d2e4e2: Pull complete 
c906e18d849e: Pull complete 
492f870e266d: Pull complete 
244116b35dd0: Pull complete 
cce597e3c2ba: Pull complete 
d91560b7b28b: Pull complete 
Digest: sha256:7d870123fe736cc912528a4f29c380d19f5efa29fc2de6b9d200cfbf92337df5
Status: Downloaded newer image for python:3.8
 ---> 7892710a667f
Step 2/6 : WORKDIR /app
 ---> Running in a8605086b745
Removing intermediate container a8605086b745
 ---> ebf1ead432b3
Step 3/6 : RUN pip3 install pypowerwall==0.5.1 bs4
 ---> Running in ff438ca084d6
Fatal Python error: pyinit_main: can't initialize time
Python runtime state: core initialized
PermissionError: [Errno 1] Operation not permitted
Current thread 0xb6fa9010 (most recent call first):
<no Python frame>
The command '/bin/sh -c pip3 install pypowerwall==0.5.1 bs4' returned a non-zero code: 1
```

This seems like the error that was in the docker logs before. Do I need to install python 3.8?

I may go for the upgrade to bullseye, but that would require me reinstalling shinobi, which I'm using to monitor a few web cams, and of course it has been working well for a year or more. lol, might be easier to buy another RPi

Let me know if you find anything else. This is beyond my expertise, but I'll keep at it. I may try it with python 3.8

**Reply by jasonacox** [reply]
Thank you, Joe!  I'm sorry this isn't as easy as it should be but your feedback is golden.  By running the alternative, I see that the real issue is indeed docker and not just the Alpine container.  That means we need to address the fundamental issue of getting the **libseccomp2** bug fixed.

## Apt-Key Fix?

I don't know why the `apt-key` command failed.  Are you behind a firewall that blocks outbound ports?  If so, you can try this instead:

```bash
sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 04EE7237B7D453EC 648ACFD622F3D138

# if it works, the rest
# Add the Buster backport repository to apt sources.list
echo 'deb http://httpredir.debian.org/debian buster-backports main contrib non-free' | sudo tee -a /etc/apt/sources.list.d/debian-backports.list

sudo apt update
sudo apt install libseccomp2 -t buster-backports

# I recommend you restart and see if that corrects the problem
```
## Manual libseccomp2 Upgrade

If the above didn't work, there is a brute force way to get the libseccomp upgrade from debian.

```bash
wget http://ftp.us.debian.org/debian/pool/main/libs/libseccomp/libseccomp2_2.4.4-1~bpo10+1_armhf.deb
sudo dpkg -i libseccomp2_2.4.4-1~bpo10+1_armhf.deb
```

## Manual Test of Proxy?

In case the above didn't work.  I am going to give you a way to run pypowerwall proxy directly.  This is for testing or temporary setup as it won't have any restart or start-on-boot function that docker enables. However, it may be useful so you can at least see the end goal. :) 

In the pypowerwall/proxy directory, create a `start.sh` file (or any name you want to use) that contains this (make sure you update the Password, Email and Host):
```bash
export PW_PASSWORD="password"
export PW_EMAIL="email@example.com"
export PW_HOST="x.x.x.x"  # IP Address of your Powerwall Gateway
export PW_TIMEZONE="America/LosAngeles"
export PW_DEBUG="no"
export PW_HTTPS="no"
export PW_PORT="8675"
export PW_CACHE_EXPIRE="10"

echo "Starting proxy at http://localhost:8675..."
python3 server.py
```
Run it with `bash start.sh`

You should now be able to go to these URLs to see pypowerwall running:

* http://localhost:8675/stats
* http://localhost:8675/ <-  This should show the animation by itself
* http://localhost:8675/soe <- battery
* http://localhost:8675/aggregates <- all system metrics
* http://localhost:8675/vitals <- all device vitals

There is a chance that telegraf will see this and start pulling in the metrics.  Speaking of which, we should verify the rest of the containers:

```bash
docker logs telegraf
docker logs influxdb
docker logs grafana
```

**Reply by maxEntropyProd** [reply]
Good news @jasonacox the second attempt at the Apt-key fixed worked! My router doesn't block outbound traffic, but the hkp:// url worked (not sure what that url is).

I did not bother to reboot, so I just reopened Powerwall Power Flow and I'm now seeing live connection to the Tesla gateway! I'm not seeing much other data, but I'm guessing that's because the powerwalls are off and the system is in standby (I'm still waiting for Tesla to fix the inverter problem...argh, but that's a different problem :)

Thank you so much for your help, as there is no way I would have been able to figure that out. Very cool!

cheers,
-joe

update: I am seeing most of the data (sting voltage, powerwall temps, home and line freq., etc, it just took a bit to start populating the graphs)

**Reply by jasonacox** [reply]
That's great news, Joe!  Good job!!!  Thanks for the update. 👏 

I'm going to add a note to the instructions (or perhaps even some code in `setup.sh`) to help better handle installations on the Raspbian GNU/Linux 10 (buster) systems like yours.  I am surprised that this hasn't come up sooner.  Buster is legacy, but not that old. Thanks for helping me work through the problem and come up with a solution.

If you are interested, grab a screenshot of your dashboard and paste it here so we can celebrate with you. 🎉 

PS - Since you have system troubles right now, you might check out the alerts that your Powerwall systems is publishing.  It sounds like Tesla knows about it, but you mind find it interesting: http://localhost:8675/alerts

**Reply by maxEntropyProd** [reply]
Great. Ya, it seems I set up the RPi4 not that long ago (but time can run slow), so hopefully your good sleuthing will help others. 

The graphics are great! here's the top bit, but so much more
![topBit](https://user-images.githubusercontent.com/16806730/181144846-f8272a35-97d5-419d-9be3-0ef452999a68.jpg)

Didn't see the alerts, just,
[
    "SystemConnectedToGrid",
    "SystemShutdown",
    "FWUpdateSucceeded",
    "SiteMaxPowerLimited",
    "PanelMaxCurrentLimited",
    "GridCodesWriteError",
    "GridCodesWriteError",
    "DeviceShutdownRequested",
    "PVAC_a014_PVS_disabled_relay",
    "PVS_a021_RapidShutdown"
]

but it's in standby. Latest new is that it's the backup switch, as they are sending a replacement for that.
Again, thanks for your help!

**Reply by jasonacox** [reply]
It looks great, Joe.  Thanks for pasting those alerts!  There are several I haven't seen before.  I log all the alerts here: https://github.com/jasonacox/pypowerwall#devices-and-alerts 

Since your system is in `SystemShutdown` you could also check the device vitals: http://localhost:8675/vitals - It contains your serial numbers so you probably don't want to paste it but you might find it interesting.
